// schema/migration.js
var dbConfig = require('./dbConfig').config,
sqladapter = require('../sqladapter/sqladapter'),
versions = [

	
//user
	{
		in: function(){
			return sqladapter.createTable(
				dbConfig.database,
				'user',
				{
					id:'INTEGER(10) AUTO_INCREMENT',
					account:'VARCHAR(255)',
					password:'TEXT',
					created_at:'TIMESTAMP NOT NULL DEFAULT now()',
					updated_at:'DATETIME NOT NULL'
				},
				'id'
			);
		},
		out: function(){
			return sqladapter.dropTable(dbConfig.database,'user');
		}
	},
//article
	{
		in: function(){
			return sqladapter.createTable(
				dbConfig.database,
				'article',
				{
					id:'INTEGER(10) AUTO_INCREMENT',
					title:'TEXT',
					body:'TEXT',
					category_id:'INTEGER',
					created_at:'TIMESTAMP NOT NULL DEFAULT now()',
					updated_at:'DATETIME NOT NULL'
				},
				'id'
			);
		},
		out: function(){
			return sqladapter.dropTable(dbConfig.database,'article');
		}
	},
//category
	{
		in: function(){
			return sqladapter.createTable(
				dbConfig.database,
				'category',
				{
					id:'INTEGER(10) AUTO_INCREMENT',
					resource_id:'INTEGER',
					resource_type_id:'INTEGER',
					name:'TEXT',
					created_at:'TIMESTAMP NOT NULL DEFAULT now()',
					updated_at:'DATETIME NOT NULL'
				},
				'id'
			);
		},
		out: function(){
			return sqladapter.dropTable(dbConfig.database,'category');
		}
	},
//resource_type
	{
		in: function(){
			return sqladapter.createTable(
				dbConfig.database,
				'resource_type',
				{
					id:'INTEGER(10) AUTO_INCREMENT',
					name:'TEXT',
					created_at:'TIMESTAMP NOT NULL DEFAULT now()',
					updated_at:'DATETIME NOT NULL'
				},
				'id'
			);
		},
		out: function(){
			return sqladapter.dropTable(dbConfig.database,'resource_type');
		}
	},
	{
		in: function(){
			return sqladapter.addColumn(dbConfig.database,'article','user_id','INTEGER(10)');
		},
		out: function(){
			return sqladapter.dropColumn(dbConfig.database,'article','user_id');
		}
	},
	{
		in: function(){
			return sqladapter.addColumn(dbConfig.database,'article','is_public','INTEGER(1) DEFAULT 0');
		},
		out: function(){
			return sqladapter.dropColumn(dbConfig.database,'article','is_public');
		}
	}
];

exports.update = function(){
	var fs = require('fs');
	fs.readFile('./node_modules/schema/version',function(err,data){
		var version = parseInt(data),i=version;
		function up(){
			fs.writeFile('./node_modules/schema/version',i);
			if (i < versions.length) {
				var c = versions[i].in();
				i++;
				c.on('end',up);
			};
		}
		up();
	});
};

exports.reverse = function(version){
	var fs = require('fs');
	var version = parseInt(data),i=version;
	function down(){
		fs.writeFile('./node_modules/schema/version',i);
		if (i > 0) {
			var c = versions[i].out();
			i--;
			c.on('end',down);
		};
	}
	down();
	
};
